ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpDataHelper.java

```diff
@@ -0,0 +1,589 @@
+import com.github.ambry.clustermap.PartitionId;
+import com.github.ambry.commons.BlobId;
+import com.github.ambry.messageformat.BlobData;
+import com.github.ambry.messageformat.BlobProperties;
+import com.github.ambry.messageformat.MessageFormatException;
+import com.github.ambry.messageformat.MessageFormatRecord;
+import com.github.ambry.store.FindToken;
+import com.github.ambry.store.FindTokenFactory;
+import com.github.ambry.store.IndexValue;
+import com.github.ambry.store.StoreKey;
+import com.github.ambry.store.StoreKeyFactory;
+import com.github.ambry.utils.SystemTime;
+import com.github.ambry.utils.Utils;
+import java.io.DataInputStream;
+import java.io.EOFException;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.RandomAccessFile;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+
+
+/**
+ * Helper class to assist in dumping data from index or log files
+ */
+public class DumpDataHelper {
+
+  ClusterMap _clusterMap;
+  String outFile;
+  FileWriter fileWriter;
+
+  public DumpDataHelper(ClusterMap clusterMap) {
+    this._clusterMap = clusterMap;
+  }
+
+  /**
+   * Initializes the output file
+   * @param outFile
+   */
+  void init(String outFile) {
+    try {
+      if (outFile != null) {
+        this.outFile = outFile;
+        fileWriter = new FileWriter(new File(outFile));
+      }
+    } catch (IOException IOException) {
+      System.out.println("IOException while trying to create File " + this.outFile);
+    }
+  }
+
+  /**
+   * Initializes the output file and the {@link FileWriter} for redirecting the output
+   * @param outFile
+   * @param fileWriter
+   */
+  void initializeOutFiles(String outFile, FileWriter fileWriter) {
+    this.outFile = outFile;
+    this.fileWriter = fileWriter;
+  }
+
+  /**
+   * Dumps all blobs in an index file
+   * @param indexFileToDump the index file that needs to be parsed
+   * @param blobList List of blobIds to be filtered for
+   * @param blobIdToMessageMap {@link HashMap} of BlobId to {@link IndexRecord} to hold the information
+   *                                          about blobs in the index after parsing
+   * @param avoidMiscLogging {@code true} if miscellaneous logging should be avoided, {@code false} otherwise
+   * @return the total number of keys/records processed
+   */
+  long dumpBlobsFromIndex(File indexFileToDump, ArrayList<String> blobList,
+      ConcurrentHashMap<String, IndexRecord> blobIdToMessageMap, boolean avoidMiscLogging) {
+    long numberOfKeysProcessed = 0;
+    try {
+      DataInputStream stream = new DataInputStream(new FileInputStream(indexFileToDump));
+      short version = stream.readShort();
+      if (!avoidMiscLogging) {
+        logOutput("version " + version);
+      }
+      if (version == 0) {
```

Comment:
the only problem I have with this is that changing versions of the index segment (or any other index code) requires changes in tools too.

Once you move these files over to the store package, I recommend reusing `IndexSegment`. The `getEntriesSince()` function might help you achieve what you want without having to rewrite all of that code.